### PR TITLE
Read/write mutex

### PR DESCRIPTION
**Pull request by jenswi-linaro**
Adds read/write mutex as needed by https://github.com/OP-TEE/optee_os/pull/1928

### CODE REVIEW

File: core/arch/arm/kernel/mutex.c [hunk 1]
```diff
@@ -91,37 +91,118 @@ static void __mutex_unlock(struct mutex *m, const char *fname, int lineno)
...
-	wq_wake_one(&m->wq, m, fname, lineno);
+	wq_wake_next(&m->wq, m, fname, lineno);
 }
 
 static bool __mutex_trylock(struct mutex *m, const char *fname __unused,
 			int lineno __unused)
 {
 	uint32_t old_itr_status;
-	enum mutex_value old_value;
+	bool can_lock;
```
Comment (lorc) [inline core/arch/arm/kernel/mutex.c]: `can_lock_write` or similar will be more descriptive. I'm looking from new developer PoV.

File: core/arch/arm/kernel/mutex.c [hunk 2]
```diff
@@ -91,37 +91,118 @@ static void __mutex_unlock(struct mutex *m, const char *fname, int lineno)
...
-	return old_value == MUTEX_VALUE_UNLOCKED;
+	return can_lock;
+}
+
+static void __mutex_read_unlock(struct mutex *m, const char *fname, int lineno)
+{
+	uint32_t old_itr_status;
+	short new_state;
+
+	old_itr_status = thread_mask_exceptions(THREAD_EXCP_ALL);
```
Comment (lorc) [inline core/arch/arm/kernel/mutex.c]: Are you deliberately not using `cpu_spin_lock_xsave()` ?

File: core/arch/arm/kernel/mutex.c [hunk 3]
```diff
@@ -91,37 +91,118 @@ static void __mutex_unlock(struct mutex *m, const char *fname, int lineno)
...
+	}
+}
+
+static bool __mutex_read_trylock(struct mutex *m, const char *fname __unused,
+				 int lineno __unused)
+{
+	uint32_t old_itr_status;
+	bool can_lock;
+
+	old_itr_status = thread_mask_exceptions(THREAD_EXCP_ALL);
```
Comment (lorc) [inline core/arch/arm/kernel/mutex.c]: The same about `cpu_spin_lock_xsave`

**Comment by jenswi-linaro** [comment]
Added missing "Signed-off-by" for "core: extend mutex to support read locks"

File: core/arch/arm/include/kernel/pseudo_ta.h [hunk 4]
```diff
@@ -34,12 +34,12 @@
...
 #include <util.h>
 
-#define PTA_MANDATORY_FLAGS	(TA_FLAG_SINGLE_INSTANCE | \
-				TA_FLAG_MULTI_SESSION | \
+#define PTA_MANDATORY_FLAGS	(TA_FLAG_MULTI_SESSION | \
 				TA_FLAG_INSTANCE_KEEP_ALIVE)
 
 #define PTA_ALLOWED_FLAGS	(PTA_MANDATORY_FLAGS | \
-				TA_FLAG_SECURE_DATA_PATH)
+				 TA_FLAG_SINGLE_INSTANCE | \
```
Comment (etienne-lms) [inline core/arch/arm/include/kernel/pseudo_ta.h]: What would it mean to enable `TA_FLAG_SINGLE_INSTANCE` in a pTA ?
pTAs are single instance by construction, even if the flag is not set.

Removing `TA_FLAG_SINGLE_INSTANCE` will also disable the effect of `TA_FLAG_MULTI_SESSION` for the pTA which will now accept only 1 active session.

To get multithreaded pTA, maybe we should rather change `tee_ta_try_set_busy()` to not lock the single instance TA mutex on pTAs entry.

File: core/arch/arm/kernel/mutex.c [hunk 5]
```diff
@@ -59,19 +59,19 @@ static void __mutex_lock(struct mutex *m, const char *fname, int lineno)
...
-		if (old_value == MUTEX_VALUE_LOCKED) {
-			wq_wait_init(&m->wq, &wqe);
+		can_lock = !m->state;
+		if (!can_lock) {
+			wq_wait_init(&m->wq, &wqe, false /* wait_read */);
 			owner = m->owner_id;
 			assert(owner != thread_get_id_may_fail());
 		} else {
-			m->value = MUTEX_VALUE_LOCKED;
+			m->state = -1; /* write locked */
```
Comment (etienne-lms) [inline core/arch/arm/kernel/mutex.c]: minor: `#define MUTEX_VALUE_LOCKED/_UNLOCKED/_READ_LOCKED`   -1/0/1 ?

File: core/arch/arm/include/kernel/pseudo_ta.h (hunk 4 above)
Comment (jenswi-linaro) [reply core/arch/arm/include/kernel/pseudo_ta.h]: You're right, `TA_FLAG_SINGLE_INSTANCE` shouldn't be removed.
Perhaps we can define a new flag instead, `TA_FLAG_CONCURRENT`?
This flag would obviously only be allowed for PTAs.

**Comment by jenswi-linaro** [comment]
Added commits for `TA_FLAG_CONCURRENT`

**Comment by jenswi-linaro** [comment]
Regarding writer starvation, I guess it depends on the use case. In the first use case, protecting the TA database it will not be a problem. To address this there would be a trade-off, I think the end result will be better if we have a problem that need to be solved (preferably before it's a major problem of course).

**Comment by lorc** [comment]
`Reviewed-by: Volodymyr Babchuk <vlad.babchuk@gmail.com>`

File: core/kernel/tee_ta_manager.c [hunk 6]
```diff
@@ -116,6 +116,9 @@ static bool tee_ta_try_set_busy(struct tee_ta_ctx *ctx)
 {
 	bool rc = true;
 
+	if (ctx->flags & TA_FLAG_CONCURRENT)
+		return true;
```
Comment (etienne-lms) [inline core/kernel/tee_ta_manager.c]: This should behave has is no `single_instance_lock` is required, i.e:
```
if (ctx->busy)
	return false;
else {
	ctx->busy = true;
	return true;
}
```

File: core/arch/arm/kernel/mutex.c (hunk 5 above)
Comment (etienne-lms) [reply core/arch/arm/kernel/mutex.c]: ok.

File: core/kernel/tee_ta_manager.c (hunk 6 above)
Comment (jenswi-linaro) [reply core/kernel/tee_ta_manager.c]: `busy` is set each time a TA is entered. With what you suggest then there would be no concurrency (actually I think we would panic due to unexpected false from `tee_ta_try_set_busy()`) or am I missing something?

**Comment by jenswi-linaro** [comment]
`TA_FLAG_CONCURRENT` for normal user TAs doesn't make sense yet as there's no synchronization primitives.

File: core/kernel/tee_ta_manager.c (hunk 6 above)
Comment (etienne-lms) [reply core/kernel/tee_ta_manager.c]: Ok, I understand. In this scope, yes, `TA_FLAG_CONCURRENT` is restricted to pTAs only.

**Comment by etienne-lms** [comment]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`
Fine for me. Sorry for the delay.
The only point that troubles me is the `TA_FLAG_CONCURRENT` exposed to user TA (libutee) but not allowed to TAs. Let's resolve this later. There is already some cleanup required in TA flags (#1597).

**Comment by jenswi-linaro** [comment]
Rebased, squashed and tags applied

**Comment by jforissier** [comment]
Like Etienne I'm not really happy with `TA_FLAG_CONCURRENT`:
1. Its purpose and limitations should be clearly documented in `user_ta_header.h`. My understanding is, it allows concurrent execution of several commands (`TEEC_InvokeCommand()` in the same session (`TEEC_OpenSession()`). Is that correct? And I also understand that user TAs cannot be allowed to do that because it is against the GP specs (2.1.3 "It is not possible to execute multiple concurrent commands within a session."). So, my proposed help text for this flag: "TA can execute multiple concurrent commands within a session (pseudo-TAs only)".
2. If the flag is used by a user-mode TA, it should be rejected. I don't see this tested in this patch.

**Comment by etienne-lms** [comment]
@jforissier. Yes, TAs can't execute command concurrently. Can't we reuse this flag for "single instance concurrency" case at session opening ?

Regarding 2., it is already implicitly checked  in https://github.com/OP-TEE/optee_os/blob/master/core/arch/arm/kernel/user_ta.c#L307.

**Comment by jenswi-linaro** [comment]
`TA_FLAG_CONCURRENT` allows concurrent execution in the context, not session.

**Comment by jforissier** [comment]
So what's the difference with single instance, multiple sessions?
What is "execution in the context"? Don't you always need a session to execute a command?
(confused :confused:)

**Comment by jforissier** [comment]
Also since pTAs don't have there own address space, I think `TA_SINGLE_INSTANCE` should be mandatory (as per GP: "Denotes a Trusted Application for which all sessions opened by clients
are directed to a single TA instance.").

**Comment by jenswi-linaro** [comment]
With `TA_FLAG_CONCURRENT` each thread of execution still need its own session. It's just that instead of being blocked on `ctx->busy` being `true` that logic is skipped allowing several sessions being active in the context at the same time.

We can skip this `TA_FLAG_CONCURRENT` flag, the only drawback is that we will not have any tests targeting read/write mutex.

**Comment by jforissier** [comment]
> allowing several sessions being active in the context at the same time.

Difference with `TA_FLAG_MULTI_SESSION`, then?

**Comment by jenswi-linaro** [comment]
`TA_FLAG_MULTI_SESSION` doesn't allow sessions to execute in parallel inside the context, it only allows several open sessions to the context. The context is still single threaded (by using `ctx->busy`).

**Comment by etienne-lms** [comment]
May rename into `TA_FLAG_CONCURRENT_CONTEXT` and definitely forbidden to user TAs ?
`#define TA_FLAG_CONCURRENT_CONTEXT	(1 << 8) /* concurrent pTA execution */`

**Comment by jforissier** [comment]
Now I get it. Thanks for being patient ;)
Let me update my proposal, then: rename `TA_FLAG_CONCURRENT` to `TA_FLAG_CONCURRENT_SESSION` with the following comment: "TA instance can execute multiple sessions concurrently (pseudo-TAs only)". This is to make it more clear what is executing concurrently (i.e., sessions), and makes a nice parallel with `TA_FLAG_MULTI_SESSION` (where the sessions are allowed to exist at the same time but are not allowed to execute concurrently).

@etienne-lms just seeing your comment, I think` _SESSION` is more exact that `_CONTEXT`.

**Comment by jenswi-linaro** [comment]
OK, I'll update

**Comment by etienne-lms** [comment]
I don't like `CONCURRENT_SESSION`. It is about concurrency of execution. One could have a single session and request concurrent invocation, which shall be handle (accepted or refused/cond-waited).
Yet what ever, we can rename later.

**Comment by jenswi-linaro** [comment]
@etienne-lms, I agree with you. But I'd rather not rename it later, I preferred the name `TA_FLAG_CONCURRENT`.

**Comment by jforissier** [comment]
Agreed, let's choose a name and stick with it.

@etienne-lms sorry I don't see your point. Same session invokes are always serialized and the flag we're discussing does not affect that. 

I find `TA_FLAG_CONCURRENT` OK but unclear because it does not say what can happen concurrently. I don't think it is asking too much to require that the use case be documented clearly (what is possible to do with this flag that was not possible before?). That has to be in the comment, now it is, and the commit log too. This said back to the naming issue: I'd rather have a vague term than one we don't agree upon so feel free to switch back to `TA_FLAG_CONCURRENT`.

**Comment by jenswi-linaro** [comment]
Well the problem is `TA_FLAG_CONCURRENT_SESSION` is that it's not clear on what's concurrent with the session.
1. Can a particular session be used concurrently?
2. Or does it mean that a particular instance of a TA would accept multiple sessions where each can be used concurrently with the other sessions?

1 is the first thing that comes to mind for me.

`TA_FLAG_CONCURRENT` is also a bit vague but one thing is very clear, there's something about the TA that will be concurrent that wasn't concurrent before.

**Comment by jforissier** [comment]
I recognize that `CONCURRENT_SESSION` could be interpreted in the two ways you mentioned :(

Then let's pick your original proposal, `TA_FLAG_CONCURRENT`, because at least it isn't misleading.

Please keep the comment in `user_ta_header.h` and add an explanation to the commit log of "core: allow multithreaded pseudo TAs " -- your text above (2) is fine with me.

With that...
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

Thanks!

**Comment by jenswi-linaro** [comment]
Squashed, rebased and tag applied

**Pull request merged**